plugins/hypervisors/vmware/src/main/java/com/cloud/storage/resource/VmwareStorageProcessor.java

```diff
@@ -2778,9 +2838,25 @@ private void unmountVmfsDatastore(VmwareContext context, VmwareHypervisorHost hy
         for (Pair<ManagedObjectReference, String> host : hosts) {
             HostMO hostMO = new HostMO(context, host.first());
 
-            HostStorageSystemMO hostStorageSystemMO = hostMO.getHostStorageSystemMO();
+            if (isDatastoreMounted(dsMO, hostMO)) {
+                HostStorageSystemMO hostStorageSystemMO = hostMO.getHostStorageSystemMO();
+
+                hostStorageSystemMO.unmountVmfsVolume(getDatastoreUuid(dsMO, hostMO));
+            }
+        }
+    }
+
+    private void unmountVmfsDatastore2(VmwareContext context, VmwareHypervisorHost hyperHost, String datastoreName,
```

Comment:
out of curiosity. Why do you have two methods with the same name? I mean, I am seeing these new methods ending in 2. What is the difference with the original ones?

Reply:
It's because I couldn't name these methods the same due to erasure "issues" with Generics.

For example:

private void mountVmfsDatastore(DatastoreMO dsMO, List<Pair<ManagedObjectReference, String>> hosts) throws Exception

private void mountVmfsDatastore2(DatastoreMO dsMO, List<HostMO> hosts) throws Exception

The Java compiler makes the second parameter in each of those methods look the same, so it complained that I had two mountVmfsDatastore methods. To solve the problem, I just changed the second method's name a little bit.

By asking this question, however, you did remind me of something I had planned to do, but forgot. I wanted to have the mountVmfsDatastore make use of the mountVmfsDatastore2 method to remove redundancy. I went ahead and implemented that in my sandbox. It will be in the next commit I push for this PR. I did the same for the unmountVmfsDatastore method.

Reply:
Same story for verifyAllHostsMountedDatastore and verifyAllHostsMountedDatastore2. I plan to make verifyAllHostsMountedDatastore leverage verifyAllHostsMountedDatastore2.